File: backend/services/evConversation.js
const catalog = require('../../shared/ev-catalog.json');
const rules = require('../../shared/ev-advice.json');
const id = v => `${v.make}|${v.model}`;
const normalizeQuestion = q => String(q || '').toLowerCase().replace(/\b(suggesy|sugest|sugges|suffest)\b/g, 'suggest').replace(/[-/]/g, ' ');
function budgetFrom(text) {
  const match = text.match(new RegExp(rules.budget, 'i')) || text.trim().match(/^([\d,]+(?:\.\d+)?)\s*(k|lakh?s?|lacs?|thousand)?[!.? ]*$/i);
  if (!match) return null;
  const value = Number((match[1] || match[3]).replace(/,/g, ''));
  const unit = (match[2] || match[4] || '').toLowerCase();
  const inr = value * (/^la/.test(unit) ? 100000 : /^(k|thousand)$/.test(unit) ? 1000 : 1);
  return Number.isFinite(inr) && inr >= 1000 && inr <= 100000000 ? inr : null;
}
function subtypeFrom(text) {
  const scooter = /\bscooters?\b/i.test(text), bike = /\b(bikes?|motorcycles?)\b/i.test(text);
  return scooter === bike ? null : scooter ? 'scooter' : 'motorcycle';
}
function listIntent(question) {
  const text = normalizeQuestion(question);
  const two = /\b(two|2)\s*(wheelers?|wheels?)\b|\btwo\s*vehicles?\b|\b(scooters?|bikes?|motorcycles?)\b/.test(text)
    || (/\b2\s*vehicles?\b/.test(text) && /\b(both|which)\b/.test(text));
  const four = /\b(four|4)\s*(wheelers?|wheels?)\b|\bfour\s*vehicles?\b|\bcars?\b/.test(text)
    || (/\b4\s*vehicles?\b/.test(text) && /\b(both|which)\b/.test(text));
  const more = /\b(more|other|another|additional|next)\b|\baur\b/.test(text);
  const bareMore = /^(?:(?:please|suggest|show|give|me|some|a few|more|others?|next|names?|options?|aur|batao|dikhao)\s*)+[.!?]*$/.test(text);
  const namedList = /\b(list|names?|models|options)\b/.test(text);
  let listing = (namedList || /\b(evs|vehicles)\b/.test(text) || two || four)
    && /\b(list|names?|which|what|show|give|suggest|recommend|all|available|availbe|kaun|konsi|batao|more|other|next)\b/.test(text);
  if (more && bareMore) listing = true;
  if (/range|battery|charg|fastest|best|highest|lowest|compare/.test(text) && !namedList) listing = false;
  if (/more\s+(details|information|info)/.test(text) && !namedList) listing = false;
  const countText = text.replace(new RegExp(rules.budget, 'gi'), '').replace(/\b(two|four|2|4)\s*(wheelers?|wheels?)\b/g, '')
    .replace(/\b(two|four|2|4)\s*vehicles?\b/g, s => /\b(both|which)\b/.test(text) ? '' : s);
  const count = countText.match(/\b(\d{1,3})\b/);
  return { kind: listing ? 'list' : 'details', more, vehicle_subtype: subtypeFrom(text), explicit_category: two || four,
    category: two && four ? 'both' : two ? 'two_wheeler' : four ? 'four_wheeler' : 'both', count: count ? Math.min(100, Math.max(1, Number(count[1]))) : null };
}
function cleanCatalogContext(value) {
  if (!value || value.version !== 1 || !['both','two_wheeler','four_wheeler'].includes(value.category)) return null;
  const ids = new Set(catalog.vehicles.map(id));
  return { version: 1, category: value.category, vehicle_subtype: ['scooter','motorcycle'].includes(value.vehicle_subtype) ? value.vehicle_subtype : null,
    brands: Array.isArray(value.brands) ? value.brands.filter(s => catalog.brands.some(b => b.make === s)).slice(0,60) : [],
    budget_inr: Number.isFinite(value.budget_inr) && value.budget_inr >= 1000 && value.budget_inr <= 100000000 ? value.budget_inr : null,
    seen: Array.isArray(value.seen) ? [...new Set(value.seen.filter(s => typeof s === 'string' && ids.has(s)))].slice(0,200) : [] };
}
function contextFor(knowledge) {
  if (knowledge.catalog_context) return knowledge.catalog_context;
  if (!knowledge.vehicles?.length) return null;
  const types = [...new Set(knowledge.vehicles.map(v => v.vehicle_type))];
  return { version: 1, vehicle_subtype: knowledge.intent?.vehicle_subtype || null, category: knowledge.intent?.category || (types.length === 1 ? types[0] : 'both'),
    brands: knowledge.intent?.kind === 'recommendation' ? knowledge.intent.scope_brands || [] : [...new Set(knowledge.vehicles.map(v => v.make))],
    budget_inr: knowledge.intent?.budget_inr ?? null, seen: knowledge.vehicles.map(id) };
}
function resolveList(question, history, identify, suppliedContext) {
  const intent = listIntent(question);
  const previous = [...(Array.isArray(history) ? history : [])].reverse().filter(h => h.role === 'user');
  const referring = /\b(these|those|this companies|same|above|their|them|inki|inke|in companies)\b/i.test(question);
  if (intent.kind !== 'list' && referring && /price|cost/i.test(question) && previous.length) Object.assign(intent, listIntent(previous[0].content));
  if (intent.kind !== 'list') return null;
  let brands = identify(question).brands.map(b => b.make);
  let context = cleanCatalogContext(suppliedContext);
  if (!context && (intent.more || referring)) {
    const prior = previous.find(h => listIntent(h.content).explicit_category || identify(h.content).brands.length || /\b(ev|vehicles|models)\b/i.test(h.content));
    if (prior) {
      const seen = (history || []).filter(h=>h.role==='assistant').flatMap(h=>identify(h.content).vehicles.map(id));
      context = { version:1, category:listIntent(prior.content).category, vehicle_subtype:subtypeFrom(prior.content), brands:identify(prior.content).brands.map(b=>b.make), budget_inr:budgetFrom(prior.content), seen };
    }
  }
  if ((intent.more || referring) && context) {
    if (!brands.length) brands = context.brands;
    if (!intent.explicit_category) { intent.category = context.category; intent.vehicle_subtype = context.vehicle_subtype || null; }
  }
  const budget = budgetFrom(question) ?? (intent.more && (!intent.explicit_category || intent.category === context?.category) ? context?.budget_inr : null) ?? null;
  const seen = new Set(intent.more ? context?.seen || [] : []);
  const pool = catalog.vehicles.filter(v => (!brands.length || brands.includes(v.make))
    && (intent.category === 'both' || v.vehicle_type === intent.category)
    && (!intent.vehicle_subtype || v.vehicle_subtype === intent.vehicle_subtype)
    && v.model_status !== 'unverified-name'
    && (budget === null || (!['racing-platform','demonstration-reference','announced-reference','prebooking-closed'].includes(v.model_status)
      && (!v.price_reference || Number(v.price_reference.amount_inr) <= budget))));
  const remaining = pool.filter(v => !seen.has(id(v)));
  const limit = intent.count || 20;
  let vehicles;
  if (intent.category === 'both') {
    const two = remaining.filter(v=>v.vehicle_type==='two_wheeler');
    const four = remaining.filter(v=>v.vehicle_type==='four_wheeler');
    const selectedFour = four.slice(0, limit-Math.min(two.length,Math.ceil(limit/2)));
    vehicles = [...two.slice(0,limit-selectedFour.length),...selectedFour];
  } else vehicles = remaining.slice(0,limit);
  vehicles.forEach(v=>seen.add(id(v)));
  return { intent:{...intent,budget_inr:budget}, vehicles, total_matches:pool.length, requested_count:intent.count,
    remaining_count:remaining.length-vehicles.length,
    catalog_context:{version:1,category:intent.category,vehicle_subtype:intent.vehicle_subtype,brands,budget_inr:budget,seen:[...seen].slice(0,200)} };
}
function priceLine(v) {
  const p=v.price_reference;
  return p ? `₹${Number(p.amount_inr).toLocaleString('en-IN')} — ${p.label}; reference checked ${p.checked_at}.` : 'Price: a verified amount is not available in this catalog.';
}
function listAnswer(question, knowledge) {
  const vehicles=knowledge.vehicles || [];
  if (!vehicles.length) return knowledge.intent?.more && knowledge.total_matches > 0
    ? 'You have seen all matching models in this catalog. I will not repeat them. Try another category, brand or budget to broaden the list.'
    : 'No matching models in this reference catalog for those filters. This does not establish market unavailability. Try another brand, category or budget.';
  const lines=[`**${vehicles.length}${knowledge.intent?.more ? ' more' : ''} EV models from the reference catalog**`];
  if (knowledge.requested_count > vehicles.length) lines.push(`You requested ${knowledge.requested_count}; only ${vehicles.length} matching entries are recorded here${knowledge.intent?.more ? ' that you have not seen' : ''}.`);
  if (knowledge.intent?.budget_inr) lines.push('Budget filter uses dated advertised prices. Models without a price need a dealer quote; they are not confirmed within your budget.');
  let index=0;
  for (const [type,title] of [['two_wheeler','Two-wheelers (scooters / bikes)'],['four_wheeler','Four-wheelers (cars / SUVs)']]) {
    const group=vehicles.filter(v=>v.vehicle_type===type);if(!group.length)continue;
    lines.push('',`### ${title}`);
    for(const v of group) lines.push(`${++index}. **${v.make} ${v.model}**${v.model_status && v.model_status !== 'reference' ? ` — ${v.model_status.replace(/-/g,' ')}` : ''}${/price|cost/i.test(question) ? ` — ${priceLine(v)}` : knowledge.intent?.budget_inr && !v.price_reference ? ' — price quote needed' : ''}`);
  }
  lines.push('','These are catalog models, not confirmed local stock. Availability varies by market, city and variant.');
  if(knowledge.remaining_count > 0)lines.push(`${knowledge.remaining_count} more matching models are recorded. Ask “more names” to continue.`);
  if(/price|cost/i.test(question))lines.push('Listed prices are dated references, not current on-road quotes. Which city and variant should we check?');
  return lines.join('\n');
}
module.exports={listIntent,resolveList,listAnswer,priceLine,budgetFrom,subtypeFrom,cleanCatalogContext,contextFor};

File: rag-service/ev_conversation.py
"""Model lists, budget-safe continuation and source labels without an external AI call."""
import json
import re
from pathlib import Path
RULES = json.loads((Path(__file__).parent.parent/'shared/ev-advice.json').read_text())

def record_id(v):
    return v['make']+'|'+v['model']

def budget_from(text):
    match = re.search(RULES['budget'], text, re.I) or re.fullmatch(r'([\d,]+(?:\.\d+)?)\s*(k|lakh?s?|lacs?|thousand)?[!.? ]*', text.strip(), re.I)
    if not match: return None
    groups = match.groups() + (None, None)
    value = float((groups[0] or groups[2]).replace(',', ''))
    unit = (groups[1] or groups[3] or '').lower()
    value *= 100000 if unit.startswith('la') else 1000 if unit in ('k','thousand') else 1
    return int(value) if 1000 <= value <= 100000000 else None

def subtype_from(text):
    scooter = bool(re.search(r'\bscooters?\b', text, re.I))
    bike = bool(re.search(r'\b(bikes?|motorcycles?)\b', text, re.I))
    return None if scooter == bike else 'scooter' if scooter else 'motorcycle'

def list_intent(question):
    text = re.sub(r'[-/]', ' ', re.sub(r'\b(suggesy|sugest|sugges|suffest)\b', 'suggest', question.lower()))
    two = bool(re.search(r'\b(two|2)\s*(wheelers?|wheels?)\b|\btwo\s*vehicles?\b|\b(scooters?|bikes?|motorcycles?)\b', text) or (re.search(r'\b2\s*vehicles?\b', text) and re.search(r'\b(both|which)\b', text)))
    four = bool(re.search(r'\b(four|4)\s*(wheelers?|wheels?)\b|\bfour\s*vehicles?\b|\bcars?\b', text) or (re.search(r'\b4\s*vehicles?\b', text) and re.search(r'\b(both|which)\b', text)))
    more = bool(re.search(r'\b(more|other|another|additional|next|aur)\b', text))
    bare_more = bool(re.fullmatch(r'(?:(?:please|suggest|show|give|me|some|a few|more|others?|next|names?|options?|aur|batao|dikhao)\s*)+[.!?]*', text))
    named_list = bool(re.search(r'\b(list|names?|models|options)\b', text))
    listing = bool((named_list or re.search(r'\b(evs|vehicles)\b', text) or two or four) and re.search(r'\b(list|names?|which|what|show|give|suggest|recommend|all|available|availbe|kaun|konsi|batao|more|other|next)\b', text))
    if more and bare_more: listing = True
    if re.search(r'range|battery|charg|fastest|best|highest|lowest|compare', text) and not named_list: listing = False
    if re.search(r'more\s+(details|information|info)', text) and not named_list: listing = False
    count_text = re.sub(RULES['budget'], '', text, flags=re.I)
    count_text = re.sub(r'\b(two|four|2|4)\s*(wheelers?|wheels?)\b', '', count_text)
    if re.search(r'\b(both|which)\b', text): count_text = re.sub(r'\b(two|four|2|4)\s*vehicles?\b', '', count_text)
    count = re.search(r'\b(\d{1,3})\b', count_text)
    return dict(kind='list' if listing else 'details', more=more, vehicle_subtype=subtype_from(text), explicit_category=two or four,
                category='both' if two and four else 'two_wheeler' if two else 'four_wheeler' if four else 'both', count=min(100,max(1,int(count[1]))) if count else None)

def clean_catalog_context(value, catalog):
    if not isinstance(value,dict) or value.get('version') != 1 or value.get('category') not in ['both','two_wheeler','four_wheeler']: return None
    ids={record_id(v) for v in catalog['vehicles']}; makes={b['make'] for b in catalog['brands']}
    budget=value.get('budget_inr')
    return dict(version=1,category=value['category'],vehicle_subtype=value.get('vehicle_subtype') if value.get('vehicle_subtype') in ['scooter','motorcycle'] else None,brands=[s for s in value.get('brands',[]) if isinstance(s,str) and s in makes][:60] if isinstance(value.get('brands'),list) else [],
                budget_inr=budget if type(budget) in (int,float) and 1000<=budget<=100000000 else None,
                seen=list(dict.fromkeys(s for s in value.get('seen',[]) if isinstance(s,str) and s in ids))[:200] if isinstance(value.get('seen'),list) else [])

def context_for(knowledge):
    if knowledge.get('catalog_context'):return knowledge['catalog_context']
    vehicles=knowledge.get('vehicles',[])
    if not vehicles:return None
    intent=knowledge.get('intent',{}); types=list(dict.fromkeys(v['vehicle_type'] for v in vehicles))
    return dict(version=1,vehicle_subtype=intent.get('vehicle_subtype'),category=intent.get('category') or (types[0] if len(types)==1 else 'both'),
                brands=intent.get('scope_brands',[]) if intent.get('kind')=='recommendation' else list(dict.fromkeys(v['make'] for v in vehicles)),
                budget_inr=intent.get('budget_inr'),seen=[record_id(v) for v in vehicles])

def resolve_list(question, history, identify, catalog, supplied_context=None):
    intent=list_intent(question)
    previous=[h for h in reversed(history or []) if h.get('role')=='user']
    referring=bool(re.search(r'\b(these|those|this companies|same|above|their|them|inki|inke|in companies)\b', question,re.I))
    if intent['kind']!='list' and referring and re.search(r'price|cost',question,re.I) and previous:intent.update(list_intent(previous[0]['content']))
    if intent['kind']!='list':return None
    brands=[b['make'] for b in identify(question)[0]]
    context=clean_catalog_context(supplied_context,catalog)
    if not context and (intent['more'] or referring):
        prior=next((h for h in previous if list_intent(h['content'])['explicit_category'] or identify(h['content'])[0] or re.search(r'\b(ev|vehicles|models)\b',h['content'],re.I)),None)
        if prior:
            seen=[record_id(v) for h in (history or []) if h.get('role')=='assistant' for v in identify(h['content'])[1]]
            context=dict(version=1,category=list_intent(prior['content'])['category'],vehicle_subtype=subtype_from(prior['content']),brands=[b['make'] for b in identify(prior['content'])[0]],budget_inr=budget_from(prior['content']),seen=seen)
    if (intent['more'] or referring) and context:
        brands=brands or context['brands']
        if not intent['explicit_category']:
            intent['category']=context['category']
            intent['vehicle_subtype']=context.get('vehicle_subtype')
    budget=budget_from(question)
    if budget is None and intent['more'] and context and (not intent['explicit_category'] or intent['category']==context['category']):budget=context['budget_inr']
    seen=list(dict.fromkeys(context['seen'])) if intent['more'] and context else []
    blocked=['racing-platform','demonstration-reference','announced-reference','prebooking-closed']
    pool=[v for v in catalog['vehicles'] if (not brands or v['make'] in brands) and (intent['category']=='both' or v['vehicle_type']==intent['category'])
          and (not intent['vehicle_subtype'] or v.get('vehicle_subtype')==intent['vehicle_subtype']) and v.get('model_status')!='unverified-name' and (budget is None or (v.get('model_status') not in blocked and (not v.get('price_reference') or v['price_reference']['amount_inr']<=budget)))]
    remaining=[v for v in pool if record_id(v) not in seen];limit=intent['count'] or 20
    if intent['category']=='both':
        two=[v for v in remaining if v['vehicle_type']=='two_wheeler'];four=[v for v in remaining if v['vehicle_type']=='four_wheeler']
        selected_four=four[:limit-min(len(two),(limit+1)//2)];vehicles=two[:limit-len(selected_four)]+selected_four
    else:vehicles=remaining[:limit]
    seen=list(dict.fromkeys(seen+[record_id(v) for v in vehicles]))
    return dict(intent={**intent,'budget_inr':budget},vehicles=vehicles,total_matches=len(pool),requested_count=intent['count'],remaining_count=len(remaining)-len(vehicles),
                catalog_context=dict(version=1,category=intent['category'],vehicle_subtype=intent['vehicle_subtype'],brands=brands,budget_inr=budget,seen=seen[:200]))

def inr(amount):
    digits=str(int(amount));last=digits[-3:];prefix=digits[:-3];groups=[]
    while prefix:groups.insert(0,prefix[-2:]);prefix=prefix[:-2]
    return '₹'+','.join(groups+[last])

def price_line(vehicle):
    p=vehicle.get('price_reference')
    return f"{inr(p['amount_inr'])} — {p['label']}; reference checked {p['checked_at']}." if p else 'Price: a verified amount is not available in this catalog.'

def list_answer(question,knowledge):
    vehicles=knowledge.get('vehicles',[]);intent=knowledge.get('intent',{})
    if not vehicles:
        return 'You have seen all matching models in this catalog. I will not repeat them. Try another category, brand or budget to broaden the list.' if intent.get('more') and knowledge.get('total_matches',0)>0 else 'No matching models in this reference catalog for those filters. This does not establish market unavailability. Try another brand, category or budget.'
    extra=' more' if intent.get('more') else '';lines=[f'**{len(vehicles)}{extra} EV models from the reference catalog**']
    if (knowledge.get('requested_count') or 0)>len(vehicles):
        extra=' that you have not seen' if intent.get('more') else ''
        lines.append(f"You requested {knowledge['requested_count']}; only {len(vehicles)} matching entries are recorded here{extra}.")
    if intent.get('budget_inr'):lines.append('Budget filter uses dated advertised prices. Models without a price need a dealer quote; they are not confirmed within your budget.')
    index=0
    for kind,title in [('two_wheeler','Two-wheelers (scooters / bikes)'),('four_wheeler','Four-wheelers (cars / SUVs)')]:
        group=[v for v in vehicles if v['vehicle_type']==kind]
        if not group:continue
        lines.extend(['',f'### {title}'])
        for v in group:
            index+=1;status=' — '+v['model_status'].replace('-',' ') if v.get('model_status') and v['model_status']!='reference' else ''
            price=' — '+price_line(v) if re.search(r'price|cost',question,re.I) else ' — price quote needed' if intent.get('budget_inr') and not v.get('price_reference') else ''
            lines.append(f"{index}. **{v['make']} {v['model']}**{status}{price}")
    lines.extend(['','These are catalog models, not confirmed local stock. Availability varies by market, city and variant.'])
    if knowledge.get('remaining_count',0)>0:lines.append(f"{knowledge['remaining_count']} more matching models are recorded. Ask “more names” to continue.")
    if re.search(r'price|cost',question,re.I):lines.append('Listed prices are dated references, not current on-road quotes. Which city and variant should we check?')
    return '\n'.join(lines)
